### ISSUE TITLE
Diffusion Models are Evolutionary Algorithms

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11889]
Diffusion Models are Evolutionary Algorithms<br>
<br>
https://ift.tt/J6boGzX<br>
<br>
Grigory Sapunov<br>
<br>

<div><div><div dir="auto">
<p><strong>Authors:</strong><span> </span><em><span>Yanbo Zhang, Benedikt Hartl, Hananel Hazan, Michael Levin</span><br></em><strong>Paper:</strong><span> </span><a href="https://arxiv.org/abs/2410.02543">https://arxiv.org/abs/2410.02543 </a><br><strong>Code:</strong><span> </span><a href="https://github.com/Zhangyanbo/diffusion-evolution">https://github.com/Zhangyanbo/diffusion-evolution </a></p>
<p>An interesting paper from Michael Levin and co. This is the same Michael Levin who has long been working on bioelectricity, artificial life, and many other biology-related topics. He gave a fantastic talk titled "What Bodies Think About: Bioelectric Computation Outside the Nervous System" at NeurIPS 2018:</p>
<p>He also did an intriguing keynote "Robot Cancer: what the bioelectrics of embryogenesis and regeneration can teach us about unconventional computing, cognition, and the software of life" at ALife 2020:</p>
<p><span>He co-authored an excellent work on cellular automata and differentiable morphogenesis (</span><a href="https://distill.pub/2020/growing-ca/">https://distill.pub/2020/growing-ca/</a><span>).</span></p>
<p>The paper claims that diffusion models are evolutionary algorithms. How so? Let's break it down.</p>
<p><span>In the biosphere, at least two processes are capable of generalization and creating novelty: </span><em>evolution</em><span> (a slow process of adaptation to the environment across generations of organisms through natural selection) and </span><em>learning</em><span> (a fast process allowing individuals to acquire knowledge and generalize subjective experience during their lifetime). Recently, there's been a surge in work about the connection between evolution and learning, from Hinton's classic "</span><a href="https://content.wolfram.com/sites/13/2018/02/01-3-6.pdf">How Learning Can Guide Evolution</a><span>" (1987) to Vanchurin, Wolf, Katsnelson, Koonin's "</span><a href="https://www.pnas.org/doi/10.1073/pnas.2120037119">Toward a theory of evolution as multilevel learning</a><span>" (2022) and Watson (not that one) and Levin's (that one) "</span><a href="https://journals.sagepub.com/doi/10.1177/26339137231168355">The collective intelligence of evolution and development</a><span>" (2023). The current work argues that a specific class of diffusion models, where a generative model performs sequential stochastic denoising, can be understood through an evolutionary process that performs natural selection, mutations, and reproductive isolation.</span></p>
<p>Let's recall both concepts:</p>
<p><strong>Diffusion models in simple terms</strong><span>. The forward diffusion process takes an image as input (it could be any other signal instead of an image) and sequentially adds noise step by step until it becomes completely noisy. The forward diffusion process isn't very interesting; what's interesting is the reverse process – it takes noise as input and sequentially removes it, "revealing" (creating) the image hidden behind it (like doing denoising). The forward and reverse processes can be called diffusion and denoising, respectively. I've previously discussed examples of diffusion models like DALLE 2 </span><a href="https://moocaholic.medium.com/openai-and-the-road-to-text-guided-image-generation-dall-e-clip-glide-dall-e-2-unclip-c6e28f7194ea">here</a><span>.</span></p>
<p><strong>Evolutionary algorithms in simple terms</strong><span>. Imagine we have a complex task (for example, finding the optimal shape of an airplane wing), and we create a set of random solutions – like a "population" of creatures in nature. We evaluate each solution according to specific criteria (how well it flies), "crossbreed" the best solutions (taking some parameters from one solution, some from another), occasionally "mutate" them randomly (slightly changing some parameters), and get a new "generation" of solutions. This process repeats many times, and gradually, as in natural selection, more successful variants survive and reproduce. Eventually, we get a solution that might not be perfect but is good enough for practical application. Usually, the structure of the parameter space is unknown beforehand, so the initial population often starts with a standard normal distribution. The main advantage of this approach is that it doesn't require a precise understanding of how the task works – it's enough to be able to evaluate the quality of solutions. Among popular methods are CMA-ES and PEPG (the latter, by the way, is from Schmidhuber and co., he was also actively involved with these </span><a href="https://people.idsia.ch/~juergen/evolution.html">https://people.idsia.ch/~juergen/evolution.html</a><span>), but there are many others. Some work with discrete parameter sets, some with continuous ones; here we're looking at the latter.</span></p>
<p>As you can see, both approaches involve iterative data updates and sampling new objects from complex distributions. Both have a combination of directed updates and random perturbations. These are selection+mutation in evolution's case, random noise+learned denoising in diffusion's case. This raises the question: is the mechanics of these two processes fundamentally connected, and is there a deep mathematical duality between biological evolution and generative modeling? Or is it all just analogy and vanity of vanities?</p>
<div><figure><a target="_blank" href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fab1ba144-1545-4303-9960-da2b53f90c3e_1221x523.png"><div></div></a></figure></div>
<p><span>First, the authors analyze evolution from the perspective of generative models. Looking at species populations in the biosphere, the variational evolutionary process can be understood as a </span><strong>transformation of distributions</strong><span> (will it be dist2dist analogous to seq2seq?), distributions of genotypes and phenotypes. Mutations and selection jointly change the shapes of these distributions. Many biologically inspired evolutionary algorithms can be understood similarly: they optimize an objective function by maintaining and iteratively changing the distribution of a large population. And this same concept – transformation of distributions – is central to many generative models: VAEs, GANs, and diffusion models learn to transform simple distributions (often standard Gaussian) into more complex ones, where samples represent meaningful images, sounds, and texts.</span></p>
<p>On the other hand, diffusion can also be viewed from an evolutionary perspective. During training, data points are noised, and the model learns to predict this added noise to reverse the process (by the way, has anyone worked on a diffusion time machine yet?). In the sampling phase, the model starts with points from a Gaussian distribution and incrementally updates them through denoising, where noise-free samples are the ideal. In this case, directed denoising can be interpreted as directed selection, and each step adds a small noise (with a minus sign?) analogous to mutations. This resembles an evolutionary process and aligns with ideas interpreting the genome as a latent space parameterization of a multi-scale generative morphogenetic process, rather than just a blueprint of an organism. If an evolutionary process is reversed, the evolved population of highly correlated and highly fit individuals will gradually dissolve, similar to the forward diffusion process.</p>
<p><span>Analogous to energy and probability in statistical physics, evolutionary tasks can be connected to generative ones by mapping fitness to probability density: high fitness corresponds to high probability density. The authors ultimately mathematically derive a new algorithm called </span><strong>Diffusion Evolution</strong><span> – an evolutionary optimization procedure based on iterative error correction similar to diffusion models, but not relying on neural networks.</span></p>
<div><figure><a target="_blank" href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fe9ea29fa-3110-4d6b-a0ee-f8f3a3b00630_1217x598.png"><div></div></a></figure></div>
<p>Here are its key features:</p>
<ol>
<li><p>Start with a population of random solutions (like noise in diffusion models)</p></li>
<li>
<p>At each step:</p>
<ul>
<li><p>Each solution is evaluated by a fitness function</p></li>
<li><p>For each solution, its "denoised" version is estimated through weighted averaging with neighboring solutions (higher weight for more successful neighbors)</p></li>
<li><p>The solution takes a small step toward its "denoised version" and receives a small random mutation</p></li>
</ul>
</li>
<li>
<p>As it progresses:</p>
<ul>
<li><p>The neighbor search radius gradually decreases (like noise reduction in diffusion models)</p></li>
<li><p>This allows exploring the solution space globally at first, then optimizing locally</p></li>
</ul>
</li>
</ol>
<p>The key advantage of the algorithm is that it can find and maintain multiple good solutions simultaneously, unlike many classical evolutionary algorithms, which usually converge to a single solution.</p>
<div><figure><a target="_blank" href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2F45b09ec6-bf94-4028-b133-d9312ceb31c3_1205x887.png"><div></div></a></figure></div>
<p>Several experiments were conducted with the new algorithm.</p>
<p>In the first experiment, they used five different two-dimensional fitness landscapes: Rosenbrock and Beale with one optimum, and Himmelblau, Ackley, and Rastrigin with multiple optima. These were compared with other evolutionary strategies: CMA-ES, OpenES, and PEPG.</p>
<div><figure><a target="_blank" href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fbb4cd23e-13a1-46ff-8217-2016a8aa72e1_1206x1187.png"><div></div></a></figure></div>
<p>Evolution was run 100 times for each method. Each experiment had a population size of 512 and used 25 iterations (except for OpenES, which needed 1000 steps to converge). Diffusion Evolution finds quality and diverse solutions, especially on the last three landscapes where other methods struggle and tend to converge to a single solution.</p>
<div><figure><a target="_blank" href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2F885e039a-8ca2-4c96-9e3c-186b9e49464d_1208x303.png"><div></div></a></figure></div>
<p>In evolutionary algorithms, fitness evaluation is often the most computationally expensive operation, so the authors tried to reduce the number of iterations by borrowing cosine scheduling from work on diffusion models. This significantly reduced the number of fitness evaluations.</p>
<p><span>In the second experiment, they proposed </span><strong>Latent Space Diffusion Evolution</strong><span>, inspired by latent space diffusion models (</span><a href="https://arxiv.org/abs/2112.10752">https://arxiv.org/abs/2112.10752</a><span>). It allows solving problems with high-dimensional parameter spaces by exploring a low-dimensional latent space. Here, the method was applied to RL tasks where a network had to learn to control the classic cart-pole system. A two-layer network with 58 parameters was used for control. Direct Diffusion Evolution works poorly, but switching to a latent space with two parameters works well. As I understand it, the transformation is performed through a random projection matrix, and it's only used to calculate distances between solutions, while the solutions themselves are updated in the original space. The result is good and works with larger networks too (they also tested it on a three-layer network with 17,410 parameters).</span></p>
<div><figure><a target="_blank" href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fcd89bea0-0219-4841-a5e1-c0cf62c288a7_1207x1061.png"><div></div></a></figure></div>
<p><span>Overall, it's a success. They also showed that working solutions can be transferred from other areas (like they transferred the idea from latent diffusion models). This approach is similar to what Tri Dao and Albert Gu use in their SSMs, when they unite SSM and something known like linear transformer into one class and transfer ideas that work on this transformer to SSM, as was done in the Mamba-2 paper (we discussed it </span><a href="https://gonzoml.substack.com/p/mamba-2-is-here">here</a><span>), for example.</span></p>
<p><span>This is all very exciting movement, showing that learning and evolution are essentially doing the same thing. And remembering the work comparing neural network training through SGD with the diffusion process (Neural Network Diffusion, </span><a href="https://gonzoml.substack.com/p/neural-network-diffusion">a beautiful paper discussed here</a><span>), then transitively we can probably say that gradient descents are also evolutionary algorithms? Are evolution and learning unifying again? And in that case, is the thermodynamic computer (we touched it </span><a href="https://gonzoml.substack.com/p/unconventional-values">here</a><span> and </span><a href="https://gonzoml.substack.com/p/thermodynamic-ai-is-getting-hotter">in more detail here</a><span>) the universal hardware for all this future AI? There's a lot to think about.</span></p>
<p>There are also open questions, for instance, a very big question about how diffusion models work in finite time, while real evolution is potentially infinite and open-ended. How can Diffusion Evolution be adapted to an open-ended setting? Could other variants of diffusion models lead to new evolutionary algorithms? (why not?) Can inductive biases from diffusion models be brought into evolutionary algorithms? How do latent diffusion models correlate with neutral genes? Can diffusion models be advanced by ideas from evolution?</p>
<p>In short, let's actively cross-pollinate!</p>
</div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via gonzoml.substack.com https://ift.tt/x6aI0Tj<br>
<br>
November 14, 2024 at 09:56AM
